File: LP/LAB_HASKELL/1_INIT/funcions_amb_nombres.hs
absValue :: Int -> Int
absValue n
    | n >=  0 = n
    | otherwise = -n

power :: Int -> Int -> Int
power _ 0 = 1
power x 1 = x
power x y = x * power x (y-1)

divisors :: Int -> [Int]
divisors n = [x | x <-[1..n], n `mod` x == 0]

isPrime :: Int -> Bool
isPrime n = divisors n == [1,n]
--isPrime 2 = True
--isPrime n
--    | n <= 1 = False
--    | n > 2 = gcd 2 n == 1


slowFib :: Int -> Int
slowFib 0 = 0
slowFib 1 = 1
slowFib n = slowFib(n-1) + slowFib(n-2)


fib :: Int->(Int, Int)
fib 0 = (0,1)
fib n
    | even n = (c, d)
    | otherwise = (d, c+d)
    where 
        (a,b) = fib(div n 2)
        c = a * (b * 2 - a)
        d = a * a + b * b
        
quickFib :: Int -> Int
quickFib 0 = 0
quickFib 1 = 1
quickFib n = fst (fib n)
        



    

File: LP/LAB_PYTHON/LAB1/funcionsAmbNombres.py

def absValue(x):
    if x < 0:
        return x * (-1)
    return x


def power(x, p):
    return powerRec(x, p)


def powerIterative(x, p):
    ret = x
    while(p > 0):
        ret = ret * x
        p = p - 1
    return ret


def powerRec(x, p):
    if p == 0:
        return 1
    else:
        return (x * powerRec(x, p-1))


def isPrime(x):
    divisors = [n for n in range (1, x+1) if x % n == 0]
    return divisors == [1, x]


def slowFib(n):
    if n == 0:
        return 0
    if n == 1:
        return 1
    else:
        return slowFib(n-1) + slowFib(n-2)


def quickFib(n):
    dp = dict()
    dp[0] = 0
    dp[1] = 1
    for i in range(2, n+1):
        dp[i] = dp[i-1] + dp[i-2]
    return dp[n]
